src/domain/opendatafabric/src/identity/dataset_refs.rs

```diff
@@ -740,3 +741,126 @@ impl std::cmp::PartialOrd for DatasetRefAny {
         Some(self.cmp(other))
     }
 }
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum DatasetRefPattern {
+    Ref(DatasetRef),
+    Pattern(Option<AccountName>, DatasetNamePattern),
+}
+
+impl DatasetRefPattern {
+    pub fn match_pattern(dataset_ref: &str, pattern: &str) -> Result<bool, InvalidPatternError> {
+        Like::<false>::like(dataset_ref, pattern)
+    }
+}
+
+impl std::str::FromStr for DatasetRefPattern {
+    type Err = ParseError<Self>;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.contains('%') {
+            return match s.split_once('/') {
+                Some((account, dn)) => match DatasetNamePattern::try_from(dn) {
+                    Ok(dnp) => match AccountName::try_from(account) {
+                        Ok(an) => Ok(Self::Pattern(Some(an), dnp)),
+                        Err(_) => Err(Self::Err::new(s)),
+                    },
+                    Err(_) => Err(Self::Err::new(s)),
+                },
+                None => match DatasetNamePattern::try_from(s) {
+                    Ok(dnp) => Ok(Self::Pattern(None, dnp)),
+                    Err(_) => Err(Self::Err::new(s)),
+                },
+            };
+        }
+        match DatasetRef::from_str(s) {
+            Ok(dr) => Ok(Self::Ref(dr)),
+            Err(_) => Err(Self::Err::new(s)),
+        }
+    }
+}
+
+super::dataset_identity::impl_parse_error!(DatasetRefPattern);
+
+#[cfg(test)]
+mod tests {
```

Comment:
We currently don't use `mod tests` anywhere in the project, instead we put all tests in the separate `tests/` directory.

This is just a stylistic choice we made very early on, and I'd be OK to revisit this decision, but let's move these tests into `test/` for now and have this discussion separately.

Reply:
Not true, we do have this type of tests. They are very useful for well-isolated classes or functions that do not require catalog dependencies.

Reply:
> we do have this type of tests

That one suite of `mod tests` that was introduced in flow system ignored the unwritten convention. 99% of our tests still reside in `tests`.

Like I said I'm not against revisiting this and starting to use `mod tests` for new code, but since all other types in this module have their tests in `test_dataset_identity.rs` - I think using `mod tests` just for one of these types will be confusing.